File: app/infrastructure/onvif.py
"""ONVIF infrastructure adapter wrapping onvif-python.

Encapsulates all ONVIF wire-protocol calls (connection checks, snapshot
and stream URI resolution, RTSP auth URL building) so services depend on
this adapter rather than the third-party client directly.
"""

from loguru import logger
from onvif import ONVIFClient
from typing import Optional


class ONVIFCameraClient:
    """Adapter that performs ONVIF operations against a remote camera.

    Each method opens a short-lived ONVIFClient connection, so instances
    are stateless and safe to share as a singleton injected into
    SnapshotService.
    """

    def test_connection(self, host: str, port: int, username: str, password: str) -> tuple[bool, list[str], str | None]:
        """Verify ONVIF connectivity and collect media profile tokens.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port (commonly 80 or 8000).
            username: The camera authentication username.
            password: The camera authentication password.

        Returns:
            A tuple of (success flag, profile tokens, error message).
            The error is None when the connection succeeds.
        """
        logger.debug(f"Testing ONVIF connection to {host}:{port}")
        try:
            client = ONVIFClient(host, port, username, password, timeout=10)
            device = client.devicemgmt()
            info = device.GetDeviceInformation()
            logger.debug(f"Device info for {host}:{port} — {info}")
            media = client.media()
            profiles = media.GetProfiles()
            profile_tokens = [p.token for p in profiles]
            logger.info(f"ONVIF connection OK to {host}:{port} ({len(profile_tokens)} profiles)")
            return True, profile_tokens, None
        except Exception as e:
            logger.warning(f"ONVIF connection failed to {host}:{port} — {e}")
            return False, [], str(e)

    def get_snapshot_uri(
        self, host: str, port: int, username: str, password: str, profile_token: str | None = None
    ) -> tuple[str | None, str | None]:
        """Resolve the JPEG snapshot URI for a camera profile.

        When profile_token is None, auto-selects the first available media
        profile before requesting the snapshot URI.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port.
            username: The camera authentication username.
            password: The camera authentication password.
            profile_token: Optional media profile token; auto-selected when None.

        Returns:
            A tuple of (snapshot URI, error message). The URI is None on
            failure and the error is None on success.
        """
        logger.debug(f"Getting snapshot URI from {host}:{port} (profile: {profile_token})")
        try:
            client = ONVIFClient(host, port, username, password, timeout=15)
            media = client.media()
            if profile_token is None:
                profiles = media.GetProfiles()
                if not profiles:
                    logger.warning(f"No media profiles found for {host}:{port}")
                    return None, "No media profiles found"
                profile_token = profiles[0].token
                logger.debug(f"Auto-selected profile {profile_token} for {host}:{port}")
            snapshot = media.GetSnapshotUri(ProfileToken=profile_token)
            uri = snapshot.Uri
            if isinstance(uri, dict):
                uri = uri.get("Uri", "")
            logger.debug(f"Snapshot URI resolved: {uri}")
            return str(uri), None
        except Exception as e:
            logger.error(f"Failed to get snapshot URI from {host}:{port} — {e}")
            return None, str(e)

    def get_stream_uri(
        self, host: str, port: int, username: str, password: str, profile_token: str
    ) -> tuple[Optional[str], Optional[str]]:
        """Resolve the RTSP unicast stream URI for a given profile.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port.
            username: The camera authentication username.
            password: The camera authentication password.
            profile_token: The media profile token to request the stream for.

        Returns:
            A tuple of (stream URI, error message). The URI is None on
            failure and the error is None on success.
        """
        logger.debug(f"Getting stream URI from {host}:{port} (profile: {profile_token})")
        try:
            client = ONVIFClient(host, port, username, password, timeout=15)
            media = client.media()
            result = media.GetStreamUri(
                {'Stream': 'RTP-Unicast', 'Transport': {'Protocol': 'RTSP'}},
                profile_token,
            )
            uri = str(result.Uri)
            logger.debug(f"Stream URI resolved: {uri}")
            return uri, None
        except Exception as e:
            logger.warning(f"Failed to get stream URI from {host}:{port} — {e}")
            return None, str(e)

    def get_jpeg_stream_uri(
        self, host: str, port: int, username: str, password: str
    ) -> tuple[Optional[str], Optional[str], Optional[str]]:
        """Find a JPEG-encoded stream URI among available profiles.

        Iterates media profiles and returns the first whose
        VideoEncoderConfiguration uses JPEG encoding.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port.
            username: The camera authentication username.
            password: The camera authentication password.

        Returns:
            A tuple of (JPEG stream URI, profile token, error message).
            Both the URI and token are None when no JPEG profile exists.
        """
        logger.debug(f"Looking for JPEG stream on {host}:{port}")
        try:
            client = ONVIFClient(host, port, username, password, timeout=15)
            media = client.media()
            profiles = media.GetProfiles()
            for p in profiles:
                if p.VideoEncoderConfiguration and p.VideoEncoderConfiguration.Encoding == 'JPEG':
                    uri, err = self.get_stream_uri(host, port, username, password, p.token)
                    if uri:
                        logger.info(f"Found JPEG stream: {uri} (profile: {p.token})")
                        return uri, p.token, None
            return None, None, "No JPEG stream profile found"
        except Exception as e:
            logger.warning(f"Failed to find JPEG stream on {host}:{port} — {e}")
            return None, None, str(e)

    def get_first_stream_uri(
        self, host: str, port: int, username: str, password: str
    ) -> tuple[Optional[str], Optional[str]]:
        """Resolve the stream URI for the first available media profile.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port.
            username: The camera authentication username.
            password: The camera authentication password.

        Returns:
            A tuple of (stream URI, error message). The URI is None when
            no profiles exist or the request fails.
        """
        logger.debug(f"Getting first available stream URI from {host}:{port}")
        try:
            client = ONVIFClient(host, port, username, password, timeout=15)
            media = client.media()
            profiles = media.GetProfiles()
            if not profiles:
                return None, "No media profiles found"
            uri, err = self.get_stream_uri(host, port, username, password, profiles[0].token)
            if uri:
                logger.info(f"First stream: {uri} (profile: {profiles[0].token})")
            return uri, err
        except Exception as e:
            return None, str(e)

    def get_best_stream_uri(
        self, host: str, port: int, username: str, password: str
    ) -> tuple[Optional[str], Optional[str], Optional[str]]:
        """Resolve the stream URI for the highest-resolution media profile.

        Selects the profile with the largest Width*Height product and
        requests its RTSP unicast stream URI.

        Args:
            host: The camera hostname or IP address.
            port: The ONVIF service port.
            username: The camera authentication username.
            password: The camera authentication password.

        Returns:
            A tuple of (stream URI, profile token, error message). The
            URI and token are None when no profiles exist or the request
            fails.
        """
        logger.debug(f"Finding best stream URI from {host}:{port}")
        try:
            client = ONVIFClient(host, port, username, password, timeout=15)
            media = client.media()
            profiles = media.GetProfiles()
            if not profiles:
                return None, None, "No media profiles found"

            best = max(
                profiles,
                key=lambda p: (
                    p.VideoEncoderConfiguration.Resolution.Width * p.VideoEncoderConfiguration.Resolution.Height
                    if p.VideoEncoderConfiguration and p.VideoEncoderConfiguration.Resolution
                    else 0
                ),
            )
            uri, err = self.get_stream_uri(host, port, username, password, best.token)
            if uri:
                logger.info(f"Best stream: {uri} (profile: {best.token})")
            return uri, best.token, err
        except Exception as e:
            return None, None, str(e)

    def build_auth_url(self, raw_uri: str, username: str, password: str) -> str:
        """Embed credentials into a URI scheme when possible.

        Inserts username:password into the URI authority so callers can
        fetch RTSP or HTTP snapshots without separate auth handling.

        Args:
            raw_uri: The original URI to amend.
            username: The camera authentication username.
            password: The camera authentication password.

        Returns:
            The URI with embedded credentials, or the original URI when
            credentials are empty or the URI has no scheme separator.
        """
        if not raw_uri:
            return raw_uri
        if username and password:
            if "://" in raw_uri:
                scheme, rest = raw_uri.split("://", 1)
                return f"{scheme}://{username}:{password}@{rest}"
        return raw_uri

File: internal/infrastructure/onvif/onvif.go
// Package onvif wraps the use-go/onvif SOAP client behind a narrow
// adapter exposing exactly the operations the camera services need
// (device info, profiles, snapshot URI, stream URIs). Responses are
// decoded with namespace-agnostic XML structs so the adapter survives
// camera firmware differences.
package onvif

import (
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/use-go/onvif"
	"github.com/use-go/onvif/device"
	media "github.com/use-go/onvif/media"
	onvift "github.com/use-go/onvif/xsd/onvif"

	"github.com/Lmex89/home-cameras/internal/config"
)

// Client talks to a single ONVIF camera over SOAP/HTTP.
type Client struct {
	timeout time.Duration
	// httpClient overrides the per-call client (test injection; nil
	// builds a default client with the configured timeout).
	httpClient *http.Client
}

// NewClient builds a client; the timeout applies per SOAP call.
//
// Args:
//
//	timeoutSeconds: Per-call timeout in seconds.
//
// Returns:
//
//	A ready client.
func NewClient(timeoutSeconds int) *Client {
	return &Client{timeout: time.Duration(timeoutSeconds) * time.Second}
}

// NewFromConfig builds a client from application settings (15s per-call
// timeout, parity with the legacy ONVIF client default).
//
// Args:
//
//	cfg: Application configuration.
//
// Returns:
//
//	A ready client.
func NewFromConfig(cfg config.Config) *Client {
	return NewClient(15)
}

// Profile describes one media profile discovered on a camera.
type Profile struct {
	Token    string
	Width    int
	Height   int
	Encoding string
}

// TestResult is the outcome of a connection probe.
type TestResult struct {
	Reachable bool
	Profiles  []Profile
	Error     string
}

// device builds a Device for a camera endpoint with a bounded client.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials (may be empty).
//
// Returns:
//
//	The device handle (which performs GetCapabilities on creation).
func (c *Client) device(host string, port int, user, pass string) (*onvif.Device, error) {
	xaddr := fmt.Sprintf("http://%s:%d/onvif/device_service", host, port)
	httpClient := c.httpClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: c.timeout}
	}
	return onvif.NewDevice(onvif.DeviceParams{
		Xaddr:      xaddr,
		Username:   user,
		Password:   pass,
		HttpClient: httpClient,
	})
}

// soapCaller is the slice of the ONVIF device the adapter needs:
// invoking a SOAP method returns the raw HTTP response.
type soapCaller interface {
	CallMethod(method any) (*http.Response, error)
}

// call invokes a SOAP method and decodes the response body into out
// (a struct with Body>... xml tags). SOAP faults and HTTP errors are
// surfaced as Go errors.
//
// Args:
//
//	dev: The device handle.
//	req: The typed request struct (e.g. media.GetProfiles{}).
//	out: Destination for the decoded response.
//
// Returns:
//
//	Any transport, HTTP, or XML error.
func call(dev soapCaller, req any, out any) error {
	resp, err := dev.CallMethod(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("SOAP HTTP %d: %s", resp.StatusCode, truncate(string(body), 300))
	}
	dec := xml.NewDecoder(resp.Body)
	if err := dec.Decode(out); err != nil {
		return fmt.Errorf("decode SOAP response: %w", err)
	}
	return nil
}

// truncate limits a string to n characters.
//
// Args:
//
//	s: The string to truncate.
//	n: Maximum length.
//
// Returns:
//
//	The truncated string.
func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

// TestConnection verifies ONVIF reachability and lists profiles.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//
// Returns:
//
//	A TestResult (never errors; failures set Error).
func (c *Client) TestConnection(host string, port int, user, pass string) TestResult {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return TestResult{Error: err.Error()}
	}
	var body struct {
		Fault struct {
			Text string `xml:"Reason>Text"`
		} `xml:"Body>Fault"`
	}
	if err := call(dev, device.GetDeviceInformation{}, &body); err != nil {
		return TestResult{Error: err.Error()}
	}
	if body.Fault.Text != "" {
		return TestResult{Error: body.Fault.Text}
	}
	profiles, err := c.getProfiles(dev)
	if err != nil {
		return TestResult{Reachable: true, Error: err.Error()}
	}
	return TestResult{Reachable: true, Profiles: profiles}
}

// getProfiles lists the media profiles of a device.
//
// Args:
//
//	dev: The device handle.
//
// Returns:
//
//	The discovered profiles.
func (c *Client) getProfiles(dev soapCaller) ([]Profile, error) {
	var body struct {
		Profiles []struct {
			Token                     string `xml:"token,attr"`
			VideoEncoderConfiguration struct {
				Encoding   string `xml:"Encoding"`
				Resolution struct {
					Width  int `xml:"Width"`
					Height int `xml:"Height"`
				} `xml:"Resolution"`
			} `xml:"VideoEncoderConfiguration"`
		} `xml:"Body>GetProfilesResponse>Profiles"`
	}
	if err := call(dev, media.GetProfiles{}, &body); err != nil {
		return nil, err
	}
	out := make([]Profile, 0, len(body.Profiles))
	for _, p := range body.Profiles {
		out = append(out, Profile{
			Token:    p.Token,
			Width:    p.VideoEncoderConfiguration.Resolution.Width,
			Height:   p.VideoEncoderConfiguration.Resolution.Height,
			Encoding: p.VideoEncoderConfiguration.Encoding,
		})
	}
	return out, nil
}

// GetProfiles returns the media profiles of a camera.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//
// Returns:
//
//	The discovered profiles.
func (c *Client) GetProfiles(host string, port int, user, pass string) ([]Profile, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return nil, err
	}
	return c.getProfiles(dev)
}

// GetSnapshotURI resolves the JPEG snapshot URI for a profile token.
// When token is empty the first available profile is used.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//	token: Media profile token (empty = auto-select first).
//
// Returns:
//
//	The snapshot URI.
func (c *Client) GetSnapshotURI(host string, port int, user, pass, token string) (string, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return "", err
	}
	if token == "" {
		profiles, err := c.getProfiles(dev)
		if err != nil {
			return "", err
		}
		if len(profiles) == 0 {
			return "", fmt.Errorf("no media profiles found")
		}
		token = profiles[0].Token
	}
	var body struct {
		MediaUri struct {
			Uri string `xml:"Uri"`
		} `xml:"Body>GetSnapshotUriResponse>MediaUri"`
	}
	if err := call(dev, media.GetSnapshotUri{ProfileToken: onvift.ReferenceToken(token)}, &body); err != nil {
		return "", err
	}
	if body.MediaUri.Uri == "" {
		return "", fmt.Errorf("empty snapshot URI")
	}
	return body.MediaUri.Uri, nil
}

// GetStreamURI resolves the RTSP unicast stream URI for a profile.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//	token: Media profile token.
//
// Returns:
//
//	The stream URI.
func (c *Client) GetStreamURI(host string, port int, user, pass, token string) (string, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return "", err
	}
	var body struct {
		MediaUri struct {
			Uri string `xml:"Uri"`
		} `xml:"Body>GetStreamUriResponse>MediaUri"`
	}
	req := media.GetStreamUri{
		StreamSetup: onvift.StreamSetup{
			Stream:    onvift.StreamType("RTP-Unicast"),
			Transport: onvift.Transport{Protocol: onvift.TransportProtocol("RTSP")},
		},
		ProfileToken: onvift.ReferenceToken(token),
	}
	if err := call(dev, req, &body); err != nil {
		return "", err
	}
	if body.MediaUri.Uri == "" {
		return "", fmt.Errorf("empty stream URI")
	}
	return body.MediaUri.Uri, nil
}

// GetJPEGStreamURI finds the first JPEG-encoded stream URI.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//
// Returns:
//
//	The stream URI and its profile token.
func (c *Client) GetJPEGStreamURI(host string, port int, user, pass string) (string, string, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return "", "", err
	}
	profiles, err := c.getProfiles(dev)
	if err != nil {
		return "", "", err
	}
	for _, p := range profiles {
		if strings.EqualFold(p.Encoding, "JPEG") {
			uri, err := c.GetStreamURI(host, port, user, pass, p.Token)
			return uri, p.Token, err
		}
	}
	return "", "", fmt.Errorf("no JPEG stream profile found")
}

// GetFirstStreamURI resolves the stream for the first profile.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//
// Returns:
//
//	The stream URI.
func (c *Client) GetFirstStreamURI(host string, port int, user, pass string) (string, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return "", err
	}
	profiles, err := c.getProfiles(dev)
	if err != nil {
		return "", err
	}
	if len(profiles) == 0 {
		return "", fmt.Errorf("no media profiles found")
	}
	return c.GetStreamURI(host, port, user, pass, profiles[0].Token)
}

// GetBestStreamURI resolves the stream of the highest-resolution profile.
//
// Args:
//
//	host, port: Camera endpoint.
//	user, pass: ONVIF credentials.
//
// Returns:
//
//	The stream URI and its profile token.
func (c *Client) GetBestStreamURI(host string, port int, user, pass string) (string, string, error) {
	dev, err := c.device(host, port, user, pass)
	if err != nil {
		return "", "", err
	}
	profiles, err := c.getProfiles(dev)
	if err != nil {
		return "", "", err
	}
	if len(profiles) == 0 {
		return "", "", fmt.Errorf("no media profiles found")
	}
	best := profiles[0]
	for _, p := range profiles[1:] {
		if p.Width*p.Height > best.Width*best.Height {
			best = p
		}
	}
	uri, err := c.GetStreamURI(host, port, user, pass, best.Token)
	return uri, best.Token, err
}

// BuildAuthURL embeds credentials into a URI's authority when possible.
//
// Args:
//
//	rawURI: The URI to amend.
//	user, pass: Credentials to embed.
//
// Returns:
//
//	The amended URI (unchanged when credentials are empty).
func (c *Client) BuildAuthURL(rawURI, user, pass string) string {
	if rawURI == "" || user == "" || pass == "" {
		return rawURI
	}
	if idx := strings.Index(rawURI, "://"); idx >= 0 {
		scheme, rest := rawURI[:idx+3], rawURI[idx+3:]
		return scheme + user + ":" + pass + "@" + rest
	}
	return rawURI
}
